Explicit UVM prefetch API

weinbe2 opened the pull request:
This PR adds a new virtual `prefetch(QudaFieldLocation, cudaStream_t)` function to the `LatticeField` class, to be overwritten by its children class, which manually calls `cudaMemPrefetchAsync` under the hood on all device fields allocated via `pool_device_malloc`. The `QudaFieldLocation` is used to specify fetching to the host or device, and the `cudaStream_t` is an argument with default value `0` specifying the CUDA stream to use, thinking forward to overlapping memory transfer and compute for deflation. This function has also been added to the `Dirac` class, overriden as appropriate for its children, to prefetch its component fields.

The `prefetch` routine is a no-op for CPU fields. It's also a no-op for CUDA fields unless both managed memory is enabled, as well as a new environment variable `QUDA_ENABLE_MANAGED_PREFETCH`. The latter environment variable is motivated by the fact that `cudaMemPrefetchAsync` suffers an O(launch latency) overhead even if the memory is already resident in the destination.

So far the `prefetch` mechanism is only used in MG setup: it evicts the `B` fields to the host once they are done being used for the coarse construct and verify, and it explicitly migrates all gauge links and clover terms (top level and coarse) to their appropriate device to the device, avoiding swapping during the first MG solve.

This PR required a refactor of the `verify` routine to be called on a per-level basis immediately after constructing a new coarse operator. This was because verifying `|(1 - P P^dagger) B| = 0` would otherwise fetch the `B` fields back to the GPU. There were other possible refactors to avoid this issue, but I thought this was helpful because it forces the check of the coarse op before building subsequent coarse levels, which is a waste of time if the coarse op is messed up somehow.

Verified for Wilson and staggered, GPU and CPU setup.

I'll add `QUDA_ENABLE_MANAGED_PREFETCH` to the wiki once this PR is approved.

weinbe2 committed:
Added support for a UVM prefetch mechanism in all CUDA fields and in all Dirac ops. Compiles, though currently called no where.

weinbe2 committed:
Added explicit prefetch support in MG solve.

weinbe2 committed:
Added a cudaStream_t argument to prefetch, defaulting to the 0 stream.

weinbe2 committed:
Fixed a printing bug in Transfer setup. Purely cosmetic.

weinbe2 committed:
Fixed segfault caused by trying to prefetch a coarser level when on the coarsest level...

weinbe2 committed:
Fixed naive mistake.

weinbe2 committed:
Refactored verify to run for each level immediately after forming the coarse links instead of waiting until the end. Was necessary to evict B fields to the host as soon as possible.

weinbe2 committed:
Merge branch 'develop' into feature/uvm-prefetch

maddyscientist commented on include/multigrid.h:
```diff
@@ -334,12 +334,16 @@ namespace quda {
+       5. The normal operator is indeed normal: im(<x|D^\dag D|x>) < epsilon
      */
-    void verify();
+    void verify(bool recrusively = false);
```
describe the `recursively` parameter in the doxygen

maddyscientist commented on lib/clover_field.cpp:
```diff
@@ -373,6 +373,23 @@ namespace quda {
     checkCudaError();
   }
 
+  void cudaCloverField::prefetch(QudaFieldLocation mem_space, cudaStream_t stream) const
```
For clover I'm thinking we should add a specialization of `prefetch` that allows us to specify whether to fetch the clover field or its inverse as opposed to only acting on both.

maddyscientist commented on lib/cuda_color_spinor_field.cpp:
```diff
@@ -574,6 +574,21 @@ namespace quda {
+      else errorQuda("Invalid QudaFieldLocation.");
+
+      cudaMemPrefetchAsync(v, bytes, dev_id, stream);
+      if (precision == QUDA_HALF_PRECISION || precision == QUDA_QUARTER_PRECISION) cudaMemPrefetchAsync(norm, norm_bytes, dev_id, stream);
```
Can you change this to `(precision == QUDA_HALF_PRECISION || precision == QUDA_QUARTER_PRECISION) && norm_bytes > 0`.  This will allow for future optimization when doing fixed-point fields without a norm field (e.g., V fields).

maddyscientist commented on lib/dirac_coarse.cpp:
```diff
@@ -373,6 +373,14 @@ namespace quda {
 
+  void DiracCoarse::prefetch(QudaFieldLocation mem_space, cudaStream_t stream) const {
+    Dirac::prefetch(mem_space, stream);
+    if (Y_d) Y_d->prefetch(mem_space, stream);
```
Perhaps we should only fetch Yhat and not Y here?  Since Yhat is only used in the solver?

maddyscientist replied:
My comment on Clover also applies here I guess.  Only prefetch what we want according to `DiracCoarse` vs `DiracCoarsePC`.

maddyscientist commented on lib/dirac_clover.cpp:
```diff
@@ -92,6 +92,11 @@ namespace quda {
 
+  void DiracClover::prefetch(QudaFieldLocation mem_space, cudaStream_t stream) const {
+    Dirac::prefetch(mem_space, stream);
+    clover.prefetch(mem_space, stream);
```
Perhaps should specialize for preconditioned operator.  `DiracClover` should only fetch clover field, and `DiracCloverPC` operator should grab the direct and inverse field according to the type of preconditioning, e.g., both parities of inverse for symmetric preconditioned operator or one parity of each for asymmetric operator.

stevengottlieb commented:
@maddyscientist recursively is misspelled in the + line of code.

Steve

On Mon, 2020-01-27 at 14:33 -0800, maddyscientist wrote:
> @maddyscientist commented on this pull request.
> 
> 
> ______________________________________________________________________
> 
> In include/multigrid.h:
> 
> >       */
> -    void verify();
> +    void verify(bool recrusively = false);
> 
> describe the recursively parameter in the doxygen
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

maddyscientist commented:
Review done.  Looks mostly good, though see comments to be addressed.  Nice work 😄

weinbe2 committed:
Addressed PR comments. Added missing TwistedClover overload of prefetch. Added ability to only prefetch clover or inverse clover, full parity or just even/odd, depending on clover solve type.

weinbe2 commented:
@maddyscientist I've addressed your comments. In particular, I added a special version of `prefetch` for `Dirac[Twisted]Clover[PC]` which only fetches the appropriate parities of the clover field and its inverse depending on the preconditioning form.... as best as I can tell, please double check my logic.

maddyscientist commented on include/clover_field.h:
```diff
@@ -6,6 +6,14 @@
 
 namespace quda {
 
+  // Prefetch type
```
For an internal enum I wouldn't bother with the `Quda` prefix

mathiaswagner replied:
For an internal enum I would suggest to always use enum classes. No need to stick to plain old C. https://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum

mathiaswagner reviewed (changes requested):
This looks good (apart from my enum class comment)

weinbe2 committed:
Replace internal clover prefetch enum with enum class.

weinbe2 committed:
Replace staggered prolong/restrict enum with enum class while I remember...

weinbe2 committed:
Apply clang-format.

weinbe2 commented:
As requested, I replaced `enum` with `enum class` in `clover_field.h` and propagated the required changes downstream. I also remembered I had an internal `enum` in `staggered_prolong_restrict.cu`, so I made the same change there. 

Also applied a `clang-format` and did a due diligence build to make sure it goes through. Since this is the home stretch, I preemptively added `QUDA_MANAGED_MEMORY_PREFETCH` to the wiki.

weinbe2 committed:
Remved Quda/QUDA from internal enums.

mathiaswagner commented on lib/clover_field.cpp:
```diff
@@ -373,6 +373,67 @@ namespace quda {
 
+  void cudaCloverField::prefetch(QudaFieldLocation mem_space, cudaStream_t stream) const
+  {
+    prefetch(mem_space, stream, CloverPrefetchType::BOTH_CLOVER_PREFETCH_TYPE);
```
what about just naming the element `BOTH`. `Having CloverPrefetchType twice seems a bit wordy.

mathiaswagner reviewed (approved):
I am not picky about my duplicate CloverPrefetchType redundancy comment. This is nice to have not blocking.

weinbe2 commented:
Totally assumed this got merged. Any objections to me hitting merge @maddyscientist @mathiaswagner >

maddyscientist commented:
Only objection is that I hit merge first 😜
